pegasus/migrations/76_migrate_channel_sources_to_s3.rb

```diff
@@ -0,0 +1,60 @@
+require 'cdo/hip_chat'
+require 'files_api'
+
+# Migration for moving JSON.parse(row[:value])['levelSource'] and ['levelHtml'] to S3 (/v3/sources API).
+Sequel.migration do
+  up do
+    HipChat.log 'Moving <b>channel</b> sources to S3...'
+    FILE_NAME = 'main.json'
+
+    # For each channel with a level(Source|Html) and no migratedToS3 flag, create a new S3 record and set migratedToS3.
+    batch_update do |row|
+      begin
+        value = JSON.parse(row[:value])
+      rescue JSON::ParserError
+        next
+      end
+
+      if (value['levelSource'] || value['levelHtml']) && !value['migratedToS3']
+        channel = storage_encrypt_channel_id row[:storage_id], row[:id]
+        body = {
+          source: value['levelSource'],
+          html: value['levelHtml']
+        }.to_json
+        SourceBucket.new.create_or_replace channel, FILE_NAME, body
```

Comment:
Yes, I suppose the whole migration would fail. If the migration is idempotent and we are not pushing any code which depends on the migration succeeding, then I suppose failing the migration is ok. I see that we're not rescuing on mysql errors so maybe this is the norm.